### PR TITLE
Protobuf schema references support

### PR DESCRIPTION
**Pull request by jjaakola-aiven**
# About this change - What it does

Add support for protobuf schema references

References: #195 

# Why this way

Initial work is in pull request #467 
Original #467 pull request description:
```
These changes add basic support of references/dependencies for protobuf schema type.
This code contains much common code which allows adding support of references/dependencies to other types.
Limitations:
The compatibility check does not support references/dependencies.
Serialization/Deserialization of protobuf does not support references/dependencies.
```

### CODE REVIEW

**Comment by libretto** [comment]
@jjaakola-aiven do You think it is a good idea to develop two different ways of the same functionality? I just have already a workaround for compatibility(compare) functionality in the Instaclustr repository and planning to release it ASAP.

File: karapace/protobuf/proto_file_element.py [hunk 1]
```diff
@@ -99,8 +99,13 @@ def __eq__(self, other: "ProtoFileElement") -> bool:  # type: ignore
...
 
-    def compare(self, other: "ProtoFileElement", result: CompareResult) -> CompareResult:
-
+    def compare(
+        self,
+        other: "ProtoFileElement",
+        result: CompareResult,
+        self_dependency_types: Optional[List[TypeElement]] = None,
+        other_dependency_types: Optional[List[TypeElement]] = None,
+    ) -> CompareResult:
```
Comment (libretto) [inline karapace/protobuf/proto_file_element.py:108]: We have a type for dependencies... and my code which I working on already use there exactly this type. please not that Your fixes will interfere with my code logic which I just debugging. You can see it at https://github.com/instaclustr/karapace/pull/24

**Comment by jjaakola-aiven** [comment]
@libretto, this PR does not implement the same functionality in different way. Your PR and commit is the basis and my commits are on top of it. I understand this creates a conflict to your implementation, but I have to consider the total quality of the feature and therefore decided to add tests and relevant changes.

**Comment by libretto** [comment]
Better add your tests to my latest code where most bugs (including reported by You) are fixed.

**Comment by libretto** [comment]
> @libretto, this PR does not implement the same functionality in different way. Your PR and commit is the basis and my commits are on top of it. I understand this creates a conflict to your implementation, but I have to consider the total quality of the feature and therefore decided to add tests and relevant changes.

@jjaakola-aiven Actually, it did.  You did the same work which I did too You can see the code in my PR.  It adds to me more work for me for resolving conflicts and rewriting code which I did in time You did Your changes. You did something in a different way. For example, I have 2 classes one for References and one for Dependencies. I understand You like use List[Reference] instead of References class. It is ok and I will replace it (actually it is not a bad style because karapace has another example of using similar array classes, but ok). By logic, it is better to have Dependency in a separate class because dependencies are resolved references. I understand You want to improve the code style. But why You can not simply tell me to do that by pointing out what I must change? I understand You want to add more tests but why You cannot add them in separate files to avoid conflicts with my code?

**Pull request closed without merging**